cpp/include/raft/sparse/convert/detail/bitmap_to_csr.cuh

```diff
@@ -0,0 +1,281 @@
+                      const bitmap_t* bitmap,
+                      index_t num_rows,
+                      index_t num_cols,
+                      nnz_t* nnz_per_row)
+{
+  auto stream              = resource::get_cuda_stream(handle);
+  const index_t total      = num_rows * num_cols;
+  const index_t bitmap_num = raft::ceildiv(total, index_t(sizeof(bitmap_t) * 8));
+
+  int dev_id, sm_count, blocks_per_sm;
+
+  cudaGetDevice(&dev_id);
+  cudaDeviceGetAttribute(&sm_count, cudaDevAttrMultiProcessorCount, dev_id);
+  cudaOccupancyMaxActiveBlocksPerMultiprocessor(
+    &blocks_per_sm, calc_nnz_by_rows_kernel<bitmap_t, index_t, nnz_t>, calc_nnz_by_rows_tpb, 0);
+
+  index_t max_active_blocks = sm_count * blocks_per_sm;
+  auto grid = std::min(max_active_blocks, raft::ceildiv(bitmap_num, index_t(calc_nnz_by_rows_tpb)));
+  auto block = calc_nnz_by_rows_tpb;
+
+  calc_nnz_by_rows_kernel<bitmap_t, index_t, nnz_t>
+    <<<grid, block, 0, stream>>>(bitmap, num_rows, num_cols, bitmap_num, nnz_per_row);
+  RAFT_CUDA_TRY(cudaPeekAtLastError());
+}
+
+/*
+  Execute the exclusive_scan within one warp with no inter-warp communication.
+  This function calculates the exclusive prefix sum of `value` across threads within the same warp.
+  Each thread in the warp will end up with the sum of all the values of the threads with lower IDs
+  in the same warp, with the first thread always getting a sum of 0.
+*/
+template <typename value_t>
+RAFT_DEVICE_INLINE_FUNCTION value_t warp_exclusive_scan(value_t value)
+{
+  int lane_id           = threadIdx.x & 0x1f;
+  value_t shifted_value = __shfl_up_sync(0xffffffff, value, 1, warpSize);
+  if (lane_id == 0) shifted_value = 0;
+
+  value_t sum = shifted_value;
+
+  for (int i = 1; i < warpSize; i *= 2) {
+    value_t n = __shfl_up_sync(0xffffffff, sum, i, warpSize);
+    if (lane_id >= i) { sum += n; }
+  }
+  return sum;
+}
+
+// Threads per block in fill_indices_by_rows_kernel.
+static const constexpr int fill_indices_by_rows_tpb = 32;
+
+template <typename bitmap_t, typename index_t, typename nnz_t>
+RAFT_KERNEL __launch_bounds__(fill_indices_by_rows_tpb)
+  fill_indices_by_rows_kernel(const bitmap_t* bitmap,
+                              const index_t* indptr,
+                              index_t num_rows,
+                              index_t num_cols,
+                              nnz_t nnz,
+                              index_t bitmap_num,
+                              index_t* indices)
+{
+  constexpr bitmap_t FULL_MASK      = ~bitmap_t(0u);
+  constexpr bitmap_t ONE            = bitmap_t(1u);
+  constexpr index_t BITS_PER_BITMAP = sizeof(bitmap_t) * 8;
+
+  int lane_id = threadIdx.x & 0x1f;
+
+  // Ensure the HBM allocated for CSR values is sufficient to handle all non-zero bitmap bits.
+  // An assert will trigger if the allocated HBM is insufficient when `NDEBUG` isn't defined.
+  // Note: Assertion is active only if `NDEBUG` is undefined.
+  if (lane_id == 0) { assert(nnz < indptr[num_rows]); }
```

Comment:
Hi @cjnolet, as per Michael's recommendation, I'd like you to review it here. There is a trade-off for checking if the `nnz` of CSR is equal to the non-zero bits number of the bitmap. Option 1: checking outside the kernel by using the `RAFT_EXPECTS`, but we had to sync stream explicitly, which may potentially break the effort of end-users on building a pipeline to improve end-2-end performance with multi-stream in the complex industry scenarios;  Option 2: use `assert` in the kernel and reuse the `indptr[num_rows]` as the non-zero bits number of bitmap, but this solution may disable by `#define NDEBUG`. The current implementation is Option 2. Thank you!